File: leetcode/42.TrappingRainWater.go
package leetcode

import "math"

func trap(height []int) int {
	if len(height) == 0 {
		return 0
	}

	left := 0
	right := len(height) - 1
	leftMax := height[left]
	rightMax := height[right]
	result := 0

	for left < right {
		if leftMax < rightMax {
			left++
			leftMax = int(math.Max(float64(leftMax), float64(height[left])))
			result += leftMax - height[left]
		} else {
			right--
			rightMax = int(math.Max(float64(rightMax), float64(height[right])))
			result += rightMax - height[right]
		}
	}

	return result
}

File: leetcode/42.TrappingRainWater.ts
function trap(height: number[]): number {
  if (!height.length) return 0

  let left = 0
  let right = height.length - 1
  let leftMax = height[left]
  let rightMax = height[right]
  let result = 0

  while (left < right) {
    if (leftMax < rightMax) {
      left++
      leftMax = Math.max(leftMax, height[left])
      result += leftMax - height[left]
    } else {
      right--
      rightMax = Math.max(rightMax, height[right])
      result += rightMax - height[right]
    }
  }

  return result
}
console.log(trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))
